### PR TITLE
TASK: Handle missing data as a soft error during resource publishing

### PR DESCRIPTION
**Pull request by dfeyer**
This change log missing data during resource publishing. By default all notifications are logged in the system log. This change also introduce ```MessageCollector``` as the default API for resource storages and targets to provide error messages during publishing.

### CODE REVIEW

**Comment by dfeyer** [comment]
:+1: Tested on 3.0, 3.1 and master

File: TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/Notification.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use TYPO3\Flow\Annotations as Flow;
+
+/**
+ * Notification
+ */
+class Notification
```
Comment (hlubek) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/Notification.php]: Why not use the Flow Message (Error) class here?

Comment (dfeyer) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/Notification.php]: Good idea

**Comment by hlubek** [comment]
I really like to improve that part because it caused us a lot of trouble already (failed deployment / update, debugging through the database, dumping SHA1s, etc.).

Question is, if we need to implement two new classes that are concerned with logging and messages / results especially for the resource publishing. I think we can implement that in a more lightweight way by just returning an array of `TYPO3\Flow\Error\Message` or a `TYPO3\Flow\Error\Result` object from `publishCollection` (and `publishResource` respectively) without breaking the interface (yeah PHP for not forcing return types here). We could handle a `null` return value as no errors and then just print them in the command controller. Would feel easier for me.

**Comment by dfeyer** [comment]
I see other place where notification can be useful (like timing monitoring in cloud deployment, ...), and having a central class is way more easy to manage them. 

I'm OK to replace the Notification class by resusing the Flow Message class family, so we just introduce a single new class, the NotificationCollection or MessageCollector if we rename it.

**Comment by hlubek** [comment]
Yes, that would also work for me, but Notification is just too similar to the base `Message` class IMHO

**Comment by dfeyer** [comment]
:+1:

File: TYPO3.Flow/Classes/TYPO3/Flow/Command/ResourceCommandController.php [hunk 2]
```diff
@@ -86,6 +94,14 @@ public function publishCommand($collection = null)
                 $this->outputLine('Publishing resources of collection "%s"', array($collection->getName()));
                 $collection->publish();
             }
+
+            if ($this->messageCollector->hasNotification()) {
+                $this->outputLine();
+                $this->outputLine('The system has generate some notifications, please check if you can solve some issue:');
```
Comment (kdambekalns) [inline TYPO3.Flow/Classes/TYPO3/Flow/Command/ResourceCommandController.php]: make that
> The system has generated some notifications, please check if you can solve those issues:

File: TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * Notification Collector
+ *
+ * @Flow\Scope("singleton")
+ */
+class MessageCollector
+{
+    /**
+     * @var \SplObjectStorage
+     */
+    protected $backends;
```
Comment (kdambekalns) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php]: `$backends` is a weird name. This contains messages, so why not name it `$messages`?

**Comment by dfeyer** [comment]
:+1:

File: TYPO3.Flow/Classes/TYPO3/Flow/Command/ResourceCommandController.php [hunk 4]
```diff
@@ -86,6 +94,14 @@ public function publishCommand($collection = null)
                 $this->outputLine('Publishing resources of collection "%s"', array($collection->getName()));
                 $collection->publish();
             }
+
+            if ($this->messageCollector->hasNotification()) {
+                $this->outputLine();
+                $this->outputLine('The system has generate some notifications. You can check and probably fix the integrity of the resource registry by using the resource:clean command.');
```
Comment (robertlemke) [inline TYPO3.Flow/Classes/TYPO3/Flow/Command/ResourceCommandController.php]: A little typo ("generate*d*"), but maybe in general rephrase like this?

"The resources were published, but a few inconsistencies were detected. You can check and probably fix the integrity of the resource registry by using the resource:clean command."

**Comment by robertlemke** [comment]
:+1: by reading - I would have tested it, too, but somehow my local setup is pretty much broken while trying to update Neos after the Flow split while being on the train.

**Comment by dfeyer** [comment]
@robertlemke Can you test this one, it's really nice if we can merge it for 2.2

**Comment by aertmann** [comment]
:+1: by reading

**Comment by robertlemke** [comment]
Tested it, works fine and is a great improvement!

Just two thoughts:

- since it's a `MessageCollector`, wouldn't the name of the method then have to be `hasMessages()` (also note the plural)?
- are we sure enough about this so we can make it part of the API? I think it's only really useful, if also other resource storages and targets can provide error messages through that mechanism

**Comment by dfeyer** [comment]
> since it's a MessageCollector, wouldn't the name of the method then have to be hasMessages() (also note the plural)?

Agree, will push a new version later today

> are we sure enough about this so we can make it part of the API? I think it's only really useful, if also other resource storages and targets can provide error messages through that mechanism

I think yes, it can be nice to have the MessageCollector a single API to collect any kind of notification, but maybe we can discuss a little bit more how the handle Message severity, better reporting, flushing non read message to the system log, ... before making it parts of the public API. I'm fine to wait 3.3 release to make it part of the API

**Comment by kdambekalns** [comment]
We wait for @dfeyer to push the changes around the `MessageCollector` and the give it another review an merge it, I'd suggest.

**Comment by dfeyer** [comment]
Sorry for the hug delay, I lost control on my planning during the last two weeks.

@robertlemke All public method are now annotated as API, it just make sense.

Ready to merge and upmerge up to 3.1 and master

File: TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+     * @param callable $callback a callback function to process every notification
+     * @return void
+     * @api
+     */
+    public function flush(callable $callback = null)
+    {
+        foreach ($this->messages as $message) {
+            /** @var Message $message */
+            $this->messages->detach($message);
+            $this->systemLogger->log('ResourceMessageCollectorMessage: ' . $message->getMessage(), $message->getSeverity());
```
Comment (johannessteu) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php]: Shouldn't this be `RessourcePublishingMessageCollectorMessage`?

Comment (dfeyer) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php]: RessourcePublishingMessage should be OK ;)

Comment (johannessteu) [inline TYPO3.Flow/Classes/TYPO3/Flow/Resource/Publishing/MessageCollector.php]: Yeah, at least something with Publishing :-P

**Comment by johannessteu** [comment]
👍

**Pull request merged**